File: src/javascript/sorting/4.mergeSort/mergeSort.js
function merge(array1, array2) {
    let combined = [];
    let i = 0;
    let j = 0;
    while (i < array1.length && j < array2.length) {
        if (array1[i] < array2[j]) {
            combined.push(array1[i]);
            i++;
        } else {
            combined.push(array2[j]);
            j++;
        }
    }
    while (i < array1.length) {
        combined.push(array1[i]);
        i++;
    }
    while (j < array2.length) {
        combined.push(array2[j]);
        j++;
    }
    return combined;
}

function mergeSort(array) {
    if(array.length < 2)  return array;
    let mid = Math.floor(array.length/2);
    let leftArray = array.slice(0, mid);
    let rightArray = array.slice(mid);
    return merge(mergeSort(leftArray), mergeSort(rightArray));
}


let originalArray = [2, 1, 4, 3, 6, 5];
let sortedArray = mergeSort( originalArray );
console.log( "\nSorted Array:", sortedArray );

File: src/python/sorting/4.mergeSort/merge_sort.py
"""
Merge Sort Algorithm

Time Complexity: O(n log n)
Space Complexity: O(n)

Merge sort is a divide-and-conquer algorithm that divides the array
into halves, recursively sorts them, and then merges the sorted halves.
"""


def merge(array1, array2):
    """
    Merges two sorted arrays into one sorted array.
    
    Args:
        array1: First sorted array
        array2: Second sorted array
    
    Returns:
        Merged sorted array
    """
    combined = []
    i = 0
    j = 0
    
    while i < len(array1) and j < len(array2):
        if array1[i] < array2[j]:
            combined.append(array1[i])
            i += 1
        else:
            combined.append(array2[j])
            j += 1
    
    # Add remaining elements
    while i < len(array1):
        combined.append(array1[i])
        i += 1
    
    while j < len(array2):
        combined.append(array2[j])
        j += 1
    
    return combined


def merge_sort(array):
    """
    Sorts an array using merge sort algorithm.
    
    Args:
        array: List of comparable elements
    
    Returns:
        Sorted array
    """
    if len(array) < 2:
        return array
    
    mid = len(array) // 2
    left_array = array[:mid]
    right_array = array[mid:]
    
    return merge(merge_sort(left_array), merge_sort(right_array))


# Usage example
if __name__ == "__main__":
    original_array = [2, 1, 4, 3, 6, 5]
    print(f"Original: {original_array}")
    sorted_array = merge_sort(original_array)
    print(f"Sorted: {sorted_array}")
